posts/shadow-dom-and-event-propagation.md

```diff
@@ -0,0 +1,267 @@
+- **Non-bubbling and non-composed events:** `mouseenter`, `mouseleave`, `pointerenter`, `pointerleave`
+- **Bubbling and composed events:** `focusin`, `focusout`, `auxclick`, `click`, `dblclick`, `mousedown`, `mousemove`, `mouseout`, `mouseover`, `mouseup`, `wheel`, `input`, `keydown`, `keyup`, `keypress`, `touchstart`, `touchend`, `touchmove`, `pointerover`, `pointerdown`, `pointermove`, `pointerup`, `pointerout`
+
+</details>
+
+## `Event.prototype.target` and event retargeting
+
+The [`Event.prototype.target`](https://developer.mozilla.org/en-US/docs/Web/API/Event/target) property references the object from which the event was dispatched. When an event is dispatched from a DOM node, its `target` is set to the node from which the event originates.
+
+To preserve shadow DOM encapsulation and avoid leaking component internals, the **`target` is updated to the host element as events cross-shadow boundaries**. This process is called event retargeting. Let's take the same nested shadow tree example that we used before to illustrate this aspect:
+
+<event-visualizer label="Nested shadow trees" event-bubbles event-composed>
+  <template>
+    <div id="a">
+      <template shadowroot="open">
+        <div id="c">
+          <template shadowroot="open">
+            <span id="e" target></span>
+          </template>
+          <p id="d"></p>
+        </div>
+      </template>
+      <img id="b"></img>
+    </div>
+  </template>
+</event-visualizer>
+
+When the composed event propagates from `div#c` shadow root to the host element, the event `target` is set to `div#c`. In the same way, once the event propagates through `div#a` shadow boundary, the event target is set to `div#a`.
+
+One interesting side-effect of event retargeting is that once the event is done propagating the event `target` is always set to the outermost host element. In the case you are doing debouncing, the event `target` should be [cached](https://github.com/salesforce/lwc/issues/2265) otherwise the debounced method loses track of the event target.
+
+## Slotted content
+
+Shadow DOM enables to slot content into a component using the `<slot>` tag. **When an event bubbles from a slotted node, the event propagates into the shadow tree first before propagating to the host element**. This is true whether the event is `composed` or not.
+
+<event-visualizer label="Slotted content" event-bubbles>
+  <template>
+    <div id="a">
+      <template shadowroot="open">
+        <div id="c">
+          <slot id="d"></slot>
+          <img id="e"></img>
+        </div>
+      </template>
+      <p id="b" target></p>
+    </div>
+  </template>
+</event-visualizer>
+
+In the example above, `p#b` is slotted into `div#a` shadow tree. When a bubbling event is dispatched from `p#b`, instead of propagating directly to `div#a` it first propagates throw all the elements in the shadow tree. It means that this event can be intercepted by `slot#d`, `div#c`, or the shadow root before reaching `div#a`.
```

Comment:
```suggestion
In the example above, `p#b` is slotted into the `div#a` shadow tree. When a bubbling event is dispatched from `p#b`, instead of propagating directly to `div#a`, it first propagates through all the elements in the shadow tree. This means that this event can be intercepted by `slot#d`, `div#c`, or the shadow root before reaching `div#a`.
```